python/tvm/relay/quantize/quantize.py

```diff
@@ -0,0 +1,247 @@
+    current quantization scheme.
+
+    Parameters
+    ---------
+    graph: Function
+        The original graph
+    """
+    _set_conv_counter(0)  # reset counter
+    return _quantize.annotate(graph)
+
+
+def calibrate(graph, dataset=None):
+    """The calibrate procedure will try to calculate the content of
+    dom_scale, nbit, clip_min, clip_max for every `simulated_quantize`
+    operator.
+
+    Parameters
+    ---------
+    graph: Function
+        The simulation graph after annotation.
+
+    dataset: list of dict of Var -> NDArray
+        The calibration dataset.
+    """
+    def power2_scale(arr):
+        """calculate weight scale with nearest mode-2 scale"""
+        val = np.amax(np.abs(arr.asnumpy()))
+        return 2**np.math.ceil(np.math.log(val, 2)) if val > 0 else 1.0
+
+    cfg = current_qconfig()
+    const_params = {}
+    quantize_op = _op.get("simulated_quantize")
+
+    def visit_func(e):
+        if isinstance(e, _expr.Call) and e.op == quantize_op:
+            ndata, ndom_scale, nclip_min, nclip_max = e.args
+            attrs = e.attrs
+            kind = attrs.kind
+            nbit = cfg.get_nbit_by_kind(kind)
+
+            valid_bit = nbit - attrs.sign
+
+            if kind == QAnnotateKind.WEIGHT:
+                var = e.args[0]
+                assert isinstance(var, _expr.Constant), "{0}".format(var)
+                scale = power2_scale(var.data)
+            else:
+                scale = cfg.global_scale
+
+            def _make_const(val):
+                return _expr.const(val, 'float32')
+
+            valid_range = 2**valid_bit
+            const_params[ndom_scale] = _make_const(scale / valid_range)
+            const_params[nclip_min] = _make_const(- (valid_range - 1))
+            const_params[nclip_max] = _make_const((valid_range - 1))
+        return
+
+    _ir_pass.post_order_visit(graph, visit_func)
+    return _expr.bind(graph, const_params)
+
+
+def realize(graph):
+    """The realize pass will transform the simulated quantized
+    graph, which computes with float32 actually, to a real low-bit
+    integer graph. It will replace the simulated_quantize with
+    several fine-grained operators like add, multiply, and shift
+    as more as possible for performance (fusion, etc.)
+
+    Parameters
+    ---------
+    graph: Function
+        The simulated graph after calibrating.
+    """
+    return _quantize.realize(graph)
+
+
+def quantize(graph, params=None, dataset=None):
+    """ The quantization procedure.
+
+    Parameters
+    ---------
+    graph: Function
+        The original graph.
+
+    params : dict of str to NDArray
+        Input parameters to the graph that do not change
+        during inference time. Used for constant folding.
+
+    dataset: list of dict of Var -> NDArray
+        The calibration dataset.
+    """
+    with _build.build_config(opt_level=3):
+        graph = _build.optimize(graph, params)
```

Comment:
Currently, alter op layout assumes un-altered layout (e.g. NCHW). Calling optimize with opt_level=3 twice leads to error.

Reply:
need update as #2356 added a target argument to `optimize`